### ISSUE TITLE
我用oneVision数据集，微调，不知道为啥会报这个错误，提示image tokes 不匹配features

### ISSUE THREAD

**Issue by mshmoon** [QwenLM/Qwen2.5-VL#1123]
valueError: Image features and image tokens do not match: tokens: 108, features 54

**Comment by wbning** [comment]
m

**Comment by wbning** [comment]
你试下这个data_qwen.py脚本:
# -*- coding: utf-8 -*-
"""
Official Qwen‑VL fine‑tune data‑loader, patched to:
1. 兼容 PyTorch 2.1+ ‑ torch.prod 新签名；
2. 无论 grid_thw 为 list 还是 tensor，都能正确求 patch 数；
3. 避免 “Image features and image tokens do not match” 报错。
"""

import os
import copy
import json
import random
import logging
import re
import time
import math
import itertools
import ast
from dataclasses import dataclass, field
from typing import Dict, Optional, Sequence, List, Tuple
from io import BytesIO
import base64
from collections.abc import Sequence as _Sequence

import numpy as np
import torch
from torch.utils.data import Dataset
from PIL import Image
from decord import VideoReader
import transformers

from . import data_list
from .rope2d import get_rope_index_25, get_rope_index_2

# ---------- 常量 ----------
IGNORE_INDEX       = -100
IMAGE_TOKEN_INDEX  = 151655
VIDEO_TOKEN_INDEX  = 151656
DEFAULT_IMAGE_TOKEN = "<image>"
DEFAULT_VIDEO_TOKEN = "<video>"

local_rank = None


# ---------- 工具 ----------
def rank0_print(*args):
    if local_rank == 0:
        print(*args)


def read_jsonl(path: str):
    with open(path, "r", encoding="utf-8") as f:
        return [json.loads(line) for line in f]


# ---------- Qwen‑2 prompt 构造 ----------
def preprocess_qwen_2_visual(
    sources,
    tokenizer: transformers.PreTrainedTokenizer,
    grid_thw: List[int] = None,
    visual_type: str = "image",
) -> Dict:
    """
    把多模态对话数据转成 Qwen‑2 系列 chat_template 可直接喂模型的 token id。
    grid_thw : list[int]，表示每张图 / 每段视频最终合并后的 patch 数，
               用来插入对应数量的 <|image_pad|> / <|video_pad|> token。
    """
    roles = {"human": "user", "gpt": "assistant"}
    system_message = "You are a helpful assistant."

    if visual_type not in {"image", "video"}:
        raise ValueError("visual_type must be either 'image' or 'video'")

    tokenizer = copy.deepcopy(tokenizer)
    chat_template = (
        "{% for message in messages %}"
        "{{'<|im_start|>' + message['role'] + '\n' + message['content'] + '<|im_end|>' + '\n'}}"
        "{% endfor %}"
        "{% if add_generation_prompt %}{{ '<|im_start|>assistant\n' }}{% endif %}"
    )
    tokenizer.chat_template = chat_template

    visual_idx = 0
    all_input_ids, all_labels = [], []

    for source in sources:
        # 保证对话以 user 开头
        if roles.get(source[0].get("from", source[0].get("role")), None) != "user":
            source = source[1:]

        input_ids, labels = [], []

        # system
        sys_ids = tokenizer.apply_chat_template([{"role": "system", "content": system_message}])
        input_ids += sys_ids
        labels    += [IGNORE_INDEX] * len(sys_ids)

        # loop conversations
        for conv in source:
            role = roles.get(conv.get("from", conv.get("role")), conv.get("role"))
            content = conv.get("value", conv.get("content"))

            if role == "user":
                tag = f"<{visual_type}>"
                if tag in content:
                    parts, replaced = content.split(tag), []
                    for idx in range(len(parts) - 1):
                        replaced.append(parts[idx])
                        pad_cnt = grid_thw[visual_idx]   # 已经是合并后的 patch 数
                        pad_tok = f"<|{visual_type}_pad|>"
                        replacement = "<|vision_start|>" + pad_tok * pad_cnt + "<|vision_end|>"
                        replaced.append(replacement)
                        visual_idx += 1
                    replaced.append(parts[-1])
                    content = "".join(replaced)

            enc_ids = tokenizer.apply_chat_template([{"role": role, "content": content}])
            input_ids += enc_ids
            if role in {"user", "system"}:
                labels += [IGNORE_INDEX] * len(enc_ids)
            else:
                tmp = enc_ids.copy()
                tmp[:3] = [IGNORE_INDEX] * 3          # <|im_start|>assistant\n
                labels += tmp

        assert len(input_ids) == len(labels)
        all_input_ids.append(input_ids)
        all_labels.append(labels)

    return dict(
        input_ids=torch.tensor(all_input_ids, dtype=torch.long),
        labels   =torch.tensor(all_labels,   dtype=torch.long),
    )


# ---------- 数据集 ----------
class LazySupervisedDataset(Dataset):
    """Qwen‑VL SFT 训练用懒加载数据集。"""

    def __init__(self, tokenizer: transformers.PreTrainedTokenizer, data_args):
        super().__init__()

        # ---------- 数据列表 ----------
        dataset_cfgs = data_args.dataset_use.split(",")
        self.list_data_dict = []
        for cfg in data_list(dataset_cfgs):
            fmt = cfg["annotation_path"].split(".")[-1]
            anns = read_jsonl(cfg["annotation_path"]) if fmt == "jsonl" else json.load(open(cfg["annotation_path"], "r"))
            if (p := cfg.get("sampling_rate", 1.0)) < 1.0:
                anns = random.sample(anns, int(len(anns) * p))
            for ann in anns:
                ann["data_path"] = cfg["data_path"]
            self.list_data_dict.extend(anns)

        rank0_print(f"Total training samples: {len(self.list_data_dict)}")
        random.shuffle(self.list_data_dict)

        # ---------- 处理器 & 其它 ----------
        self.tokenizer  = tokenizer
        self.data_args  = data_args
        proc            = data_args.image_processor
        proc.max_pixels = data_args.max_pixels
        proc.min_pixels = data_args.min_pixels
        proc.size["longest_edge"]  = data_args.max_pixels
        proc.size["shortest_edge"] = data_args.min_pixels

        # rope 选择
        self.get_rope_index = get_rope_index_25 if data_args.model_type == "qwen2.5vl" else get_rope_index_2

        # 视频相关默认
        self.video_max_total_pixels = getattr(data_args, "video_max_total_pixels", 1664 * 28 * 28)
        self.video_min_total_pixels = getattr(data_args, "video_min_total_pixels",  256 * 28 * 28)

    # ---------- 基础信息 ----------
    def __len__(self): return len(self.list_data_dict)

    # ---------- 统一图片处理 ----------
    def process_image_unified(self, path: str):
        proc   = copy.deepcopy(self.data_args.image_processor)
        img    = Image.open(path).convert("RGB")
        res    = proc.preprocess(img, return_tensors="pt")
        return res["pixel_values"], res["image_grid_thw"][0]  # (1,C,H,W), (T,H,W)

    # ---------- 视频处理 ----------
    def process_video(self, path: str):
        vr  = VideoReader(path, num_threads=4)
        total, fps = len(vr), vr.get_avg_fps()
        sec  = total / fps
        interval, vmin, vmax = self.data_args.base_interval, self.data_args.video_min_frames, self.data_args.video_max_frames
        need = min(max(round(sec / interval), vmin), vmax)
        idx  = np.unique(np.linspace(0, total - 1, need, dtype=int))
        vid  = vr.get_batch(idx).asnumpy()

        proc = copy.deepcopy(self.data_args.image_processor)
        proc.max_pixels = self.data_args.video_max_frame_pixels
        proc.min_pixels = self.data_args.video_min_frame_pixels
        proc.size["longest_edge"]  = proc.max_pixels
        proc.size["shortest_edge"] = proc.min_pixels

        out  = proc.preprocess(images=None, videos=vid, return_tensors="pt")
        spg  = [proc.temporal_patch_size / fps] * len(out["video_grid_thw"][0])
        return out["pixel_values_videos"], out["video_grid_thw"][0], spg

    # ---------- 主入口 ----------
    def __getitem__(self, idx: int):
        # 多次尝试，防止 NAS / NFS 偶发失败
        for _ in range(3):
            try:
                return self._fetch(idx)
            except Exception as e:
                print(f"[WARN] fetch sample {idx} failed:", e)
        raise RuntimeError(f"Cannot retrieve sample {idx}.")

    # ---------- 具体流程 ----------
    def _fetch(self, idx: int) -> Dict[str, torch.Tensor]:
        item = self.list_data_dict[idx]
        conversations = item["conversations"]
        merge_size = self.data_args.image_processor.merge_size

        # ---- 图像 ----
        if "image" in item:
            img_dir = item["data_path"]
            img_lst = item["image"]
            if isinstance(img_lst, str):
                img_lst = [img_lst]

            img_tensors, img_grids = [], []
            for f in img_lst:
                t, g = self.process_image_unified(os.path.join(img_dir, f))
                img_tensors.append(t);   img_grids.append(g)

            # 计算每张图合并后的 patch 数
            grid_cnt = []
            for g in img_grids:
                g_tensor = g if torch.is_tensor(g) else torch.tensor(g)
                cnt = int(torch.prod(g_tensor).item() // (merge_size ** 2))
                grid_cnt.append(cnt)

            data = preprocess_qwen_2_visual([conversations], self.tokenizer, grid_cnt, "image")
            pos, _ = self.get_rope_index(
                merge_size, data["input_ids"], torch.stack(img_grids, 0)
            )

            sample = dict(
                input_ids     = data["input_ids"][0],
                labels        = data["labels"][0],
                position_ids  = pos,
                pixel_values  = img_tensors,
                image_grid_thw= img_grids,
            )
            return sample

        # ---- 视频 ----
        elif "video" in item:
            vid_dir = item["data_path"]
            vid_lst = item["video"]
            if isinstance(vid_lst, str):
                vid_lst = [vid_lst]

            vid_tensors, vid_grids, sec_per_grid = [], [], []
            for f in vid_lst:
                t, g, s = self.process_video(os.path.join(vid_dir, f))
                vid_tensors.append(t); vid_grids.append(g); sec_per_grid.extend(s)

            grid_cnt = []
            for g in vid_grids:
                g_tensor = g if torch.is_tensor(g) else torch.tensor(g)
                cnt = int(torch.prod(g_tensor).item() // (merge_size ** 2))
                grid_cnt.append(cnt)

            data = preprocess_qwen_2_visual([conversations], self.tokenizer, grid_cnt, "video")
            pos, _ = self.get_rope_index(
                merge_size, data["input_ids"],
                video_grid_thw=torch.stack(vid_grids, 0),
                second_per_grid_ts=sec_per_grid,
            )

            return dict(
                input_ids         = data["input_ids"][0],
                labels            = data["labels"][0],
                position_ids      = pos,
                pixel_values_videos = vid_tensors,
                video_grid_thw    = vid_grids,
            )

        # ---- 纯文本 ----
        data = preprocess_qwen_2_visual([conversations], self.tokenizer, [], "image")
        pos  = torch.arange(0, data["input_ids"].size(1)).view(1, -1).unsqueeze(0).expand(3, -1, -1)
        return dict(
            input_ids    = data["input_ids"][0],
            labels       = data["labels"][0],
            position_ids = pos,
        )


# ---------- pad & collator ----------
def pad_and_cat(tensors: List[torch.Tensor]) -> torch.Tensor:
    max_len = max(t.shape[2] for t in tensors)
    padded  = [torch.nn.functional.pad(t, (0, max_len - t.shape[2]), value=1) for t in tensors]
    return torch.cat(padded, dim=1)


@dataclass
class DataCollatorForSupervisedDataset:
    tokenizer: transformers.PreTrainedTokenizer

    def __call__(self, instances: Sequence[Dict]) -> Dict[str, torch.Tensor]:
        keys = ("input_ids", "labels", "position_ids")
        input_ids, labels, pos_ids = ([inst[k] for inst in instances] for k in keys)

        pad_id = self.tokenizer.pad_token_id
        input_ids = torch.nn.utils.rnn.pad_sequence(input_ids, batch_first=True, padding_value=pad_id)
        labels    = torch.nn.utils.rnn.pad_sequence(labels,    batch_first=True, padding_value=IGNORE_INDEX)
        pos_ids   = pad_and_cat(pos_ids)

        batch = dict(
            input_ids     = input_ids,
            labels        = labels,
            attention_mask= input_ids.ne(pad_id),
            position_ids  = pos_ids,
        )

        # 可能有图 / 视频
        if any("pixel_values" in inst for inst in instances):
            imgs  = list(itertools.chain(*(inst["pixel_values"] for inst in instances if "pixel_values" in inst)))
            grids = list(itertools.chain(*(inst["image_grid_thw"] for inst in instances if "image_grid_thw" in inst)))
            batch["pixel_values"]   = torch.cat(imgs, 0)
            batch["image_grid_thw"] = torch.stack(grids, 0)
        else:
            batch["pixel_values"] = batch["image_grid_thw"] = None

        if any("pixel_values_videos" in inst for inst in instances):
            vids  = list(itertools.chain(*(inst["pixel_values_videos"] for inst in instances if "pixel_values_videos" in inst)))
            vgrid = list(itertools.chain(*(inst["video_grid_thw"]     for inst in instances if "video_grid_thw"     in inst)))
            batch["pixel_values_videos"] = torch.cat(vids, 0)
            batch["video_grid_thw"]      = torch.stack(vgrid, 0)
        else:
            batch["pixel_values_videos"] = batch["video_grid_thw"] = None

        return batch


# -------- make data module ----------
def make_supervised_data_module(tokenizer, data_args):
    ds = LazySupervisedDataset(tokenizer, data_args)
    collator = DataCollatorForSupervisedDataset(tokenizer)
    return dict(train_dataset=ds, eval_dataset=None, data_collator=collator)


if __name__ == "__main__":
    print("⚙️  data_qwen.py standalone test passed.")

**Comment by Timothyxxx** [comment]
大概率是你max pixels设置太大，或者是图文交错的比较多，然后max tokens设置太小了，<|image_pad|>加上去之后被截断了，导致另一侧的image feature上来之后和pad对照之后不匹配了

**Comment by Natt1e** [comment]
请问为什么多个视频/图片的拼接用的是cat？
